### DISCUSSION TITLE
(MongoDB) Prisma cant build index on 1:1 self referrence

### DISCUSSION THREAD

**Discussion by laneme** [prisma/prisma#24436] [Q&A] [answered]
### Question

I have a schema like this

```prisma
model Content {
  id             String   @id @default(auto()) @map("_id") @db.ObjectId
  title          String
  altLangId       String?  @unique @db.ObjectId
  LocalArticle   Content? @relation("LocalArticle", fields: [altLangId], references: [id], onDelete: NoAction, onUpdate: NoAction)
  EnglishArticle Content? @relation("LocalArticle")
}
```

The relation part is newly introduced and works fine after `prisma generate`. But if I run `prisma db push`, i get an error

```txt
Error: MongoDB error
Kind: Command failed: Error code 11000 (DuplicateKey): Index build failed: 04e02055-18a6-4a25-b154-440bae7dfc25: Collection MYDB.Content ( d3e28aaa-f218-4cf4-b2aa-8df1161cad79 ) :: caused by :: E11000 duplicate key error collection: MYDB.Content index: Content_altLangId_key dup key: { altLangId: null }, labels: {}
   0: schema_core::state::SchemaPush
             at schema-engine/core/src/state.rs:461
```

I believe the duplicate error is due to the unique index and the missing/null `altLangId` field. But the unique index is apparently required by for self referencing schemas.

How do you solve this

### How to reproduce (optional)

Run `prisma db push` with the relations from above appeneded to a model.

### Information about Prisma Schema, Client Queries and Environment (optional)

- OS: Ubuntu 22
- Database: MongoDB v6
- Node.js version: v18, v20
- Prisma: v5

**Comment by ludralph** [comment, marked as answer]
Hi @laneme 

Thank you for raising this question.

Looking at the error message, I do think you are running into an issue where you have more than one document with a null value in its index entry, the index build fails with a duplicate key error. See this section of MongoDB [documentation](https://www.mongodb.com/docs/manual/core/index-unique/#missing-document-field-in-a-unique-single-field-index).

One way to solve this would be to create a [partial unique index](https://www.mongodb.com/docs/manual/core/index-partial/#std-label-partial-index-with-unique-constraints) directly through MongoDB Atlas.

The query to create a sparse index is
```js
db.collection.createIndex({ field: 1 }, { unique: true, sparse: true })
```
Replace `collection` with the name of your collection and `field` with the field you want to create the unique index on.

You can also take a look at the suggested [solution](https://github.com/prisma/prisma/issues/20217#issuecomment-1638596578).

**Reply by laneme** [reply]
Thank you. The `db push` does work after I create the index directly myself.

There's one caveat tho.

I have to use the same index name prisma tries assign (which is visible in the error message).
Otherwise i'll get the same error, plus the index i created is removed by prisma.

```node
// ❌ doesn't work, i get the same error
db.Content.createIndex({altLangId: 1}, {unique: 1, sparse: true })

// ✅ `Content_altLangId_key` does work (same name as the one in error)
db.Content.createIndex({altLangId: 1}, {unique: 1, sparse: true, name: "Content_altLangId_key"})
